libs/datamodel/core/src/common/preview_features.rs

```diff
@@ -1,57 +1,155 @@
+// (Usually) Append-only list of features.
+features!(
+    ConnectOrCreate,
+    TransactionApi,
+    NativeTypes,
+    GroupBy,
+    CreateMany,
+    AtomicNumberOperations,
+    AggregateApi,
+    Middlewares,
+    Distinct,
+    UncheckedScalarInputs,
+    SqlServer,
+    MongoDb,
+    OrderByRelation,
+    NApi,
+    SelectRelationCount,
+    OrderByAggregateGroup
+);
+
+// Mapping of which active, deprecated and hidden
+// features are valid in which place in the datamodel.
+lazy_static! {
+    /// Generator preview features
+    pub static ref GENERATOR: FeatureMap = {
+        FeatureMap::default().with_active(vec![
+            SqlServer,
+            OrderByRelation,
+            NApi,
+            SelectRelationCount,
+            OrderByAggregateGroup
+        ]).with_hidden(vec![
+            MongoDb
+        ]).with_deprecated(vec![
+            AtomicNumberOperations,
+            AggregateApi,
+            Middlewares,
+            NativeTypes,
+            Distinct,
+            ConnectOrCreate,
+            TransactionApi,
+            UncheckedScalarInputs,
+            GroupBy,
+            CreateMany
+        ])
+    };
+
+    /// Datasource preview features.
+    pub static ref DATASOURCE: FeatureMap = {
+        FeatureMap::default()
+    };
+}
+
+#[derive(Debug, Default)]
+pub struct FeatureMap {
+    /// Valid, visible features.
+    active: Vec<PreviewFeature>,
+
+    /// Deprecated features.
+    deprecated: Vec<PreviewFeature>,
+
+    /// Hidden preview features are valid features, but are not propagated into the tooling
+    /// (as autocomplete or similar) or into error messages (eg. showing a list of valid features).
+    hidden: Vec<PreviewFeature>,
+}
+
+impl FeatureMap {
+    pub fn active_features(&self) -> &[PreviewFeature] {
+        &self.active
+    }
+
+    pub fn hidden_features(&self) -> &[PreviewFeature] {
+        &self.hidden
+    }
+
+    pub fn with_active(mut self, active: Vec<PreviewFeature>) -> Self {
+        self.active = active;
+        self
+    }
+
+    pub fn with_hidden(mut self, hidden: Vec<PreviewFeature>) -> Self {
+        self.hidden = hidden;
+        self
+    }
+
+    pub fn with_deprecated(mut self, deprecated: Vec<PreviewFeature>) -> Self {
+        self.deprecated = deprecated;
+        self
+    }
+
+    pub fn is_valid(&self, flag: &PreviewFeature) -> bool {
+        self.active.contains(flag) || self.hidden.contains(flag)
+    }
+
+    pub fn is_deprecated(&self, flag: &PreviewFeature) -> bool {
+        self.deprecated.contains(flag)
+    }
+}
+
+impl Serialize for PreviewFeature {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+/// Lowercases first character.
+/// Assumes 1-byte characters!
+pub fn decapitalize(s: &str) -> String {
+    let first_char = s.chars().next().unwrap();
+    format!("{}{}", first_char.to_lowercase(), s[1..].to_owned())
```

Comment:
nitpick: I think the `.to_owned()` is not necessary here (maybe you need to `&s[1..]` though).

Reply:
Required by the compiler. `str` is of unknown size at compile time.

Reply:
`&str` (`&s[1..]`) should have a known size, though

Reply:
Compiler says no 🤷